### ISSUE TITLE
Feature Request: Make a Date Variable

### ISSUE THREAD

**Issue by dovy** [metabase/metabase#3575]
I would love to be able to specify a condition as variable. In the case of a Dashboard, I have a question I want to display at different time intervals.

Right now the only way to do that is to create a question for each and every time interval. I'd love to be able to use a question as a template and generate time intervals. Here's an example:

NPS rating for 1 week, 2 weeks, 3 weeks, 1 month, 3 month, 6 month.

I'd love NOT to make 3 questions for each because each and every week I'll have to add 3 new questions for the new week and I'd much rather it be automatic.  ;)

**Comment by camsaul** [comment]
I think this might be covered by #1543

**Comment by tlrobinson** [comment]
@dovy We do have dashboard filters but that applies a filter to every question (that you've wired up to the filter) on a dashboard. http://www.metabase.com/docs/latest/users-guide/07-dashboard-filters

Are you wanting to create one question and add it to a dashboard multiple times with different intervals?

**Comment by dovy** [comment]
@tlrobinson exactly.

**Comment by salsakran** [comment]
so effectively, this would mean:
a parameterized question,
the ability to set a specific dashcard's question's parameter on a per-card basis.

Interesting...

**Comment by dovy** [comment]
The use case is this:

I calculate NPS based on score. NPS is net promoter score. The calculation of the groups is: 1-6 = detractor, 7&8=neutral, and 9&10 promoters.

I want to be able to show a breakdown on chart of the detractors, promoters, and neutrals according to score based on time.

Then I can actually show the NPS score which is a diff calculation, and display that on a graph as well.

;)

**Comment by tlrobinson** [comment]
I updated the title to something that I think better reflects what you want.

Basicall you want to be able to have multiple instances of a card on a dashboard (or even within a multiseries card on a dashboard) with different fixed values for each card, rather than wiring them up to a dashboard-wide filter widget.

**Comment by aagarwal-gtr** [comment]
+1 want to be able to simultaneously parameterize all the questions (parameterized sql queries themselves) in a dashboard. use case is: 

> +1 our use case is that we have customer support team handling different customers so we want all of them to be able to go to a dashboard (or monitor it) and change filter for the customer name and then only see data relevant to that customer. schema of the dashboard will be common for everyone.

> @mazameli thanks for responding. i wasnt clear enough. so all our questions are parameterized sql queries so what they do right now is go to the question and change the parameter to get say last 6 months activity data for a certain customer. we want them to be able to access this along with some other queries (maybe all possible data they may need for a customer) from the dashboard and with maybe a line graph for this. from what i understand one way out is to change our questions to get all the data and then have the a meta filter on the dashboard but we dont want that because then we will just be handling tons of data unnecessarily. dashboard parameters will be more suitable for us than filters. does this make sense?

**Comment by bhanna** [comment]
@aagarwal-gtr We have a similar use case where I want to use this as Support dashboard. This one change would make it incredible. 

I was thinking about it as how the text variable works on an individual question, but applied to an entire dashboard. 

[Email = name@domain.com] + date range in a dashboard filter would let me build questions that do searches on a single user and see activity over time. So powerful!

**Comment by tlrobinson** [comment]
@aagarwal-gtr @bhanna Are you sure you're asking for the same thing the original issues was asking for? We already have the ability to add text and date filter widgets to a dashboard, which can be applied to every card on the dashboard (once they are wired up appropriately). http://www.metabase.com/docs/latest/users-guide/07-dashboard-filters

The original issue was asking to apply fixed values to parameters on individual instances of cards within a dashboard, for example if you wanted a dashboard that showed the same question over different (fixed) time periods. I'm not sure how that applies to the support dashboard use-case.

**Comment by tlrobinson** [comment]
@aagarwal-gtr

> from what i understand one way out is to change our questions to get all the data and then have the a meta filter on the dashboard but we dont want that because then we will just be handling tons of data unnecessarily.

I'm not sure I understand the problem. Dashboard filters apply a filter to the actual queries being executed against the database. They aren't performing a query for all the data then filtering it down at the dashboard level.

If you're concerned about support people having access to the unfiltered question, that's legitimate. For SQL queries you can set the variable to be "required" and not give it a default value. Then in the dashboard you would add a dashboard filter that you map to that question's variable. The question won't load until a value is supplied.

We don't currently have a way to require a parameter for the GUI questions.

**Comment by salsakran** [comment]
@tlrobinson I'm still wrapping my brain around how this might work, but we've also dealt with this sort of thing with eg Spot and some other internal users. 

I think the idea would be to parameterize a sql question (or use an equivalent construct for gui queries), and then *within* a table, we would provide a way to generate N cards using a parameter. Eg, the a really simple example would be a blog with 10 posts. If we wanted a dashboard that had a single card with views per day for each post we would use:

```
SELECT count(*), day
FROM post_views 
WHERE post_id={{postID}}
```

and then generate a card in a dashboard for `ids = [1,2,3,4,5,6,7,8,9,10]`

I have no idea what a good UX for this would be however, but it bears a bit of thinking, as it's come up a lot.

**Comment by salsakran** [comment]
@aagarwal-gtr + @bhanna please confirm that my interpretation of what you're asking for is accurate.

**Comment by salsakran** [comment]
and to re-iterate @tlrobinson's interpretation, we currently allow you to add dashboard parameters (eg, an ID filter widget) that you can connect to a customer ID, and end up with a dashboard where all cards in the dashboard would display the specific stats for that given customer.

that's described in http://www.metabase.com/docs/latest/users-guide/07-dashboard-filters

**Comment by tlrobinson** [comment]
To add a pseudo-code analogy, currently we allow you to do the equivalent of this:

```javascript
function CustomerDashboard(email) {
  CustomerRevenueThisMonth(email);
  CustomerReventThisYear(email);
}
```

This issue is asking for the equivalent of the following

```javascript
function CustomerDashboard(email) {
  CustomerRevenue(email, THIS_MONTH);
  CustomerRevenue(email, THIS_YEAR);
}
```

**Comment by aagarwal-gtr** [comment]
i gave this some thought and i realized that the current dashboard filters feature _does_ satisfy the customer support use case i had in mind. i am not able to verify this specifically because i dont have access to the data i had at the time, anymore. Not sure why I wasnt able to make it work. I understand what the OP here wants is different, as described by @tlrobinson. and on that i dont have much to add.

**Comment by MarkRx** [comment]
Any updates on this?

**Comment by apognu** [comment]
A secondary benefits to being able to set filters at the card-level would be the ability to create read-only shared dashboards with variable questions.

As far as I can tell, right now, if you have a dashboard filter, the value can be changed by any user that has access to the dashboard, even through the Share feature.

My use-case is to create a dashboard per customer, filtered to a view for their own data, without having to duplicate all the questions for each customer, and without them having access to the filter widget.

**Comment by theborakompanioni** [comment]
It has been a while since the last update. Any information on whether this will be considered?

In my case I want the same question displayed multiple times with different time and limit settings.. e.g.
```sql
SELECT 
  X, SUM(Y) as sum_y
FROM Z 
WHERE created_at > NOW() - INTERVAL({{var_days}} DAY)
GROUP BY X
ORDER BY sum_y DESC
LIMIT {{var_limit}}
```

Question A: `var_days = 1` and `var_limit = 3`
Question B: `var_days = 7` and `var_limit = 10`
Question C: `var_days = 30` and `var_limit = 10`

Do you think such functionality should be supported by metabase?

**Comment by 1ucian0** [comment]
I would like to be able to deactivate/activate series in a card...

**Comment by anbeier-commits** [comment]
I would like to expand this issue on collections.

The constant variables can be defined per collection and will be inherited by sub-selections.

They can be used in all questions/dashboards assigned to the collection, either directly or through a dashboard. They are either visible and adjustable by the user or they are invisible and cannot be altered.

This would be really helpful when using the same question again and again over several sites or departments or countries.

So you might use and need to manage only one question suitable for many sections/domains/areas, that are only different by some where conditions in the question/query.

What do you think?

Andreas

**Comment by flamber** [comment]
@anbeier-commits Please open a new feature request and reference this, since Collections are very different from Dashboards.

**Comment by alekseyp** [comment]
My use case would be to create new card for one of the group by values.

I do have ~100 objects with multiple lines each, when they all are on one graph it's just impossible to see anything. So I would like to automatically create 100 cards on dashboard and just scroll through them.

**Comment by shakudu** [comment]
I work with a lab enviroment that would like to track different elements in samples.
My dashboard would have a date-filter. But i have to create one question for every single tracked element instead of reusing the same question with a set variable for the element shown. I could do Dashboard-filters for all of them but GUI-wise it would look horrible.

**Comment by codingjoe** [comment]
Just a precaution, should this ever be implemented: Please triple check for SQL injection attacks. Allowing SQL syntax from user input is shouting SQL injection. It's no. 3 on the OWASP list for a good reason 😏

**Comment by sfaut** [comment]
An use case :
- I manage products, each product has 1 category
- I want to display on dasboard 3 cards, with 1 category's stats per card
- Actually I must write 3 questions (1 question per category I want to include on dashboard), I would prefer write 1 question and create 3 dashboard cards with category as parameter, something like the "Default filter widget value" on question editor

Thx !

**Comment by steven1u** [comment]
Hello there! Any update on this?

Another workaround would be to create 1 question, replicate it into 3 cards in the dashboard, then create 3 filters with different default value, each filter linked to a card

**Comment by bfelbo** [comment]
Yes, this is the workaround we use. However, it litters the top of the dashboard with filters that you shouldn't touch 😅

**Comment by calops** [comment]
I'm confused about the fact that this feature has never been worked on. Right now the alternatives are:

* Create multiple versions of the same question and use them in the dashboard. This works, but then it becomes a nightmare to update the general query as it needs to be updated the same way in multiple places.
* Create multiple filters on the dashboard and link them to their respective cards. This works, but it clutters the top of the dashboard with editable filters that actually shouldn't be edited at all.

**Comment by Xusifob** [comment]
I agree with @calops , it is a complicated workaround that makes dashboards very inefficient

Is there a way to maybe develop this part ? I could spend a bit of time looking at if someone can guide me to it

**Comment by ljmachado** [comment]
Another work around is to reference the question as a CTE, apply the filter as sql and save as a new question.
This still has the issue that needs a new question and you have to rebuild the viz, but it's less code replication at least.
This feature should be an easy win that would have long lasting effects for dashboard management.

**Comment by anbeier-commits** [comment]
I am still seeking a solution as I am looking for

I'd like to add another option that would work:

Just make it possible to user (some important) metabase fields within the query as variables, e.g.

report_card.id
report_dashboardcard.dashboard_id
collection.location

This would be sufficient for using the same question over and over again with an external mapping table. e.g.

`select .... from table t
left outer join mapping_table m on m.dashboard_id= {report_dashboardcard.dashboard_id}
where t.company = coalesce(m.company_id,-1)`

**Comment by JGrubb** [comment]
Please 🙏  - this would be such an incredible win for maintainability if the same question could be reused for viz'ing various segments (passed as filter values in the question settings) without duplicating and changing 1 filter value on each.

**Comment by antoine-b-smartway** [comment]
+1 for the need, it would really help for maintainability. I'm forced to dupplicate SQL queries so many time because of this.
My needs are the same than described before : put several times the same question on a single dashboard with an hardcoded parameter for each question.

**Comment by pvblivs** [comment]
+1

One more use case: A dashboard displaying the development of key metrics over time for all cities you operate in in separate charts.

Really doesn't sound complicated, just means it needs to be possible to (pre)select a filter on the dashboard card, instead of globally on the dashboard - technically probably also super simple, you could for instance overwrite the default selection.

Tableau has this kind of stuff for decades, so easy to say: "This chart, but for all 30 cities", but I'd be totally fine creating a card for all 30 cities manually - just not creating the global filter mess :-/

**Comment by ranquild** [comment]
You can do this in v51 with time grouping parameters https://github.com/metabase/metabase/issues/42118. To have the same timeseries question displayed with different time granularities, do the following:
1. Add the question multiple times to the dashboard.
2. Add "time grouping" parameters. One parameter per time granularity you want to get:
![Image](https://github.com/user-attachments/assets/1c7ffba9-b93b-4032-b330-650eb0484a6c)
3. Connect each parameter one card and set the parameter default value to the desired granularity:
![Image](https://github.com/user-attachments/assets/d6e0edf2-d5e0-47f1-9d97-6e60f66380d7)
4. Change card titles to reflect the time granularities:
![Image](https://github.com/user-attachments/assets/ef999961-9213-4901-9f58-e5a6992a33b2)
![Image](https://github.com/user-attachments/assets/f13a2406-548d-49cb-9920-b079e74c479c)
5. Save the dashboard and see the desired result:
![Image](https://github.com/user-attachments/assets/50729300-7b46-4b75-855d-90a845cf4388)

**Comment by calops** [comment]
That's still kind of a hack. having parameters for things that are intended to be statically configured is not only poor UX, it's also error-prone as users can edit the parameters and have the views not represent what the titles are saying.

And let's be honest, this in particular should make it obvious that this is merely a workaround for a need that isn't covered properly:
![Image](https://github.com/user-attachments/assets/1e1fdd97-6bd9-494d-b3fa-938944f5c74c)

**Comment by ranquild** [comment]
@calops yes, we're thinking about optimizing this case. Still, the workaround exists now and it didn't exit before.

**Comment by canay-f** [comment]
hi, is there any update on using parameters besides time grouping?

for instance, i want to create a "compare to" parameter like 1D, 7D, 30D and use the selected value inside expressions:
`Offset(Sum([Spend After Tax]), [Compare Days])`

this would make period comparisons much easier without creating separate columns for each option.
